ic-canister/ic-canister-macros/src/canister_call.rs

```diff
@@ -21,36 +23,108 @@ impl Parse for CanisterCall {
+    let args = normalize_args(&input.method_call.args);
+    let cdk_call = get_cdk_call(
+        quote! {#canister.principal()},
+        &method_name,
+        &args,
+        &input.response_type,
+    );
+
+    let expanded = quote! {
+        {
+            #[cfg(target_arch = "wasm32")]
+            {
+                #cdk_call
+            }
+
+            #[cfg(not(target_arch = "wasm32"))]
+            {
+                #canister.#inner_method(#args)
+            }
+        }
+    };
+
+    TokenStream::from(expanded)
+}
+
+struct VirtualCanisterCall {
+    principal: Expr,
+    method_name: LitStr,
+    args: ExprTuple,
+    response_type: Type,
+}
+
+impl Parse for VirtualCanisterCall {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
+        let principal = input.parse()?;
+        input.parse::<Token![,]>()?;
+
+        let method_name = input.parse()?;
+        input.parse::<Token![,]>()?;
+
+        let args = input.parse()?;
+
+        input.parse::<Token![,]>()?;
+        let response_type = input.parse()?;
+
+        Ok(Self {
+            principal,
+            method_name,
+            args,
+            response_type,
+        })
     }
+}
+
+pub(crate) fn virtual_canister_call(input: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(input as VirtualCanisterCall);
+    let principal = &input.principal;
+    let args = normalize_args(&input.args.elems);
+    let method_name = input.method_name.value();
+    let response_type = &input.response_type;
+
+    let cdk_call = get_cdk_call(quote! {#principal}, &method_name, &args, response_type);
 
-    let response_type = input.response_type;
     let is_tuple = matches!(response_type, Type::Tuple(_));
 
-    let cdk_call = if is_tuple {
-        quote! {
-            ::ic_cdk::api::call::call::<_, #response_type>(#canister.principal(), #method_name, (#args))
-        }
+    let (decode, tuple_index) = if is_tuple {
+        (
+            quote! { ::ic_cdk::export::candid::decode_args::<#response_type>(&result) },
+            quote! {},
+        )
     } else {
-        let mut elems = Punctuated::new();
-        elems.push_value(response_type);
-        elems.push_punct(Default::default());
-        let tuple_response_type = Type::Tuple(TypeTuple {
-            paren_token: Default::default(),
-            elems,
-        });
+        (
+            quote! { ::ic_cdk::export::candid::decode_args::<(#response_type,)>(&result) },
+            quote! {.0},
+        )
+    };
 
-        quote! {
-            async {::ic_cdk::api::call::call::<_, #tuple_response_type>(#canister.principal(), #method_name, (#args)).await.map(|x| x.0)}
+    let responder_call = quote! {
+        async {
+            let encoded_args = match ::ic_cdk::export::candid::encode_args((#args)) {
+                Ok(v) => v,
+                Err(e) => return Err((::ic_cdk::api::call::RejectionCode::Unknown, format!("failed to serialize arguments: {}", e))),
```

Comment:
This syntax cannot be used inside macro:
```
error: there is no argument named `e`
  --> src/token/src/canister/is20_notify.rs:72:5
   |
72 |     virtual_canister_call!(tx.to, "transaction_notification", (notification,), ()).await
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: did you intend to capture a variable `e` from the surrounding scope?
   = note: to avoid ambiguity, `format_args!` cannot capture variables when the format string is expanded from a macro
   = note: this error originates in the macro `virtual_canister_call` (in Nightly builds, run with -Z macro-backtrace for more info)
```